flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -411,8 +410,7 @@ private static boolean isNotMarkedForDeletion(Stat stat) {
                         final RetrievableStateHandle<T> stateHandle = getAndLock(path);
                         stateHandles.add(new Tuple2<>(stateHandle, path));
                     } catch (NotExistException ignored) {
-                        // Concurrent deletion, retry
-                        continue retry;
+                        // entry is subject for deletion and can be ignored
```

Comment:
Just a note for understanding the previous behavior.

This either means that the node is marked for deletion (we can't acquire the lock) or that we're it has been already deleted (concurrent deletion).

The intention of the `goto retry` was to fail fast here as we already knew the `cversion` of the root node has changed. This assumption no longer holds as it could have been simply marked for deletion. If there is a concurrent deletion happening, we'll still be able to catch that later on by the `cversion` check.

Reply:
Maybe it would be nice to expand the comment along these lines, it's rather difficult to understand why this could be ignored.